std/typecons.d

```diff
@@ -5394,6 +5394,582 @@ private
     extern(C) pure nothrow Object typecons_d_toObject(void* p);
 }
 
+// Begin addition
+
+private template ContravariantRhsT(Rhs)
+{
+    static if (is (Rhs == class))
+    {
+        alias ContravariantRhsT = const Object;
+    }
+    else static if (is (Rhs == struct))
+    {
+        alias ContravariantRhsT = auto ref const Unqual!Rhs;
```

Comment:
Unfortunately, I don't think you can use storage classes like `auto ref`, `ref`, `lazy` or `scope` with aliases.

Reply:
interesting. If these attributes are accepted but "noop" then their presence should be prevented by the compiler.

Reply:
> Unfortunately, I don't think you can use storage classes like `auto ref`, `ref`, `lazy` or `scope` with aliases.

It didn't even occur to me that this might be ignored. I agree with @Basile-z that this should throw a compiler error if it's not supported: silently ignoring can lead to hard to discover bugs.

@ZombineDev do you know what is the reason to ignore the storage classes?

Reply:
Because they are not type constructors, but parameter storage classes and `alias` can "point" to only types, symbols and occasionally values of which storage classes are neither.
For example, you can't declare a function local or class/struct member variable with the same storage classes as a parameter.

This is unfortunately by design and will take an exceptional DIP to convince W&A otherwise. Manu has tried to argue for making `ref` a first class citizen (i.e. a type constructor), so it can play nice with `std.meta` metaprogramming for years in the newsgroup, but has not succeeded so far. Many others have tried to independently argue that `scope` should be a transitive (again meaning a type constructor like `const`), but have again failed.

For example, my interpretation of the outcome of this bug report https://issues.dlang.org/show_bug.cgi?id=17764 is that Walter values parameter storage classes not being type constructors more then memory safety.

I wonder what would be @atilaneves take on this.

Reply:
I don't have an opinion on it yet, but would love to hear arguments for/against/about it.

Reply:
In general, the problem with parameter storage classes is that they don't compose well. (Other storage classes like `static` are fine as they are, because you rarely have composition problems there.)

In case of DIP1000 / `scope` the problem is more dramatic as the compiler simply gives up and allows memory-unsafe code in `@safe` functions.

Reply:
> Manu has tried to argue for making ref a first class citizen

about `ref`, as you should know, it has been partially addressed lately (`ref` for alias function is handled but still not as a type ctor, rather as a special case).

Reply:
> ref for alias function is handled but still not as a type ctor, rather as a special case

@Basile-z thanks for point this out! Can you point me to the PR that changed this lately?

Reply:
https://dlang.org/changelog/2.086.0.html#funcliteralsbyref

Reply:
I know this enhancement, but it's unrelated. The core issue is the following:
```D
alias RI = ref int;
pragma (msg, RI);   // prints "int" during compilation

void inc(RI ri) { ri++; }

void main()
{
    int x = 10;
    x.writeln;   // prints "10" at run-time
    x.inc;
    x.writeln;   // prints "10" at run-time
}
```

That doesn't work either:
```D
alias Ref(T) = ref T;
void inc(Ref!int ri) { ri++; }
pragma (msg, typeof(&inc));   // prints "void function(int ri)" at compile-time
```